File: plugin/src/main/java/org/beangle/maven/plugin/patch/PatchWarMojo.java
package org.beangle.maven.plugin.patch;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import org.apache.maven.artifact.Artifact;
import org.apache.maven.plugin.AbstractMojo;
import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugin.MojoFailureException;
import org.apache.maven.plugins.annotations.Component;
import org.apache.maven.plugins.annotations.LifecyclePhase;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.ResolutionScope;
import org.apache.maven.project.MavenProject;
import org.apache.maven.settings.Settings;

@Mojo(name = "patch-hibernate-war", defaultPhase = LifecyclePhase.PREPARE_PACKAGE, requiresDependencyCollection = ResolutionScope.COMPILE_PLUS_RUNTIME)
public class PatchWarMojo extends AbstractMojo {

  @Component
  private Settings settings;

  @Component
  private MavenProject project;

  @Override
  public void execute() throws MojoExecutionException, MojoFailureException {
    if (!project.getPackaging().equals("war")) {
      getLog().info("Hibernate war patching supports war project.Skip jar/pom projects.");
      return;
    }
    boolean findBeangleJpa = false;
    boolean findHibernate = false;
    for (Artifact artifact : project.getArtifacts()) {
      if (artifact.getArtifactId().equals("beangle-data-jpa")) findBeangleJpa = true;
      if (artifact.getArtifactId().equals("hibernate-core")) findHibernate = true;
    }
    if (!(findHibernate && findBeangleJpa)) return;
    try {
      for (String patch : Patches.files) {
        String target = project.getBuild().getOutputDirectory() + patch;
        InputStream is = getClass().getResourceAsStream("/patches" + patch + "file");
        getLog().info("Patching " + patch);
        new File(target.substring(0, target.lastIndexOf("/"))).mkdirs();
        new File(target).createNewFile();
        copy(is, new FileOutputStream(target));
      }
    } catch (Exception e) {
      e.printStackTrace();
    } finally {
    }
  }

  static long copy(InputStream source, OutputStream sink) throws IOException {
    long nread = 0L;
    byte[] buf = new byte[8192];
    int n;
    while ((n = source.read(buf)) > 0) {
      sink.write(buf, 0, n);
      nread += n;
    }
    source.close();
    sink.close();
    return nread;
  }

  public Settings getSettings() {
    return settings;
  }

  public void setSettings(Settings settings) {
    this.settings = settings;
  }

}

File: plugin/src/main/scala/org/beangle/maven/plugin/patch/PatchWarMojo.scala
/*
 * Beangle, Agile Development Scaffold and Toolkit
 *
 * Copyright (c) 2005-2017, Beangle Software.
 *
 * Beangle is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Beangle is distributed in the hope that it will be useful.
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Beangle.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.beangle.maven.plugin.patch

import java.io.{ File, FileOutputStream }
import scala.collection.JavaConverters.asScalaSet
import org.apache.maven.plugin.AbstractMojo
import org.apache.maven.plugins.annotations.{ Component, Mojo, LifecyclePhase, ResolutionScope, Parameter }
import org.apache.maven.project.MavenProject
import org.apache.maven.settings.Settings
import org.beangle.commons.io.IOs

@Mojo(name = "patch-hibernate-war", defaultPhase = LifecyclePhase.PREPARE_PACKAGE, requiresDependencyCollection = ResolutionScope.COMPILE_PLUS_RUNTIME)
class PatchWarMojo extends AbstractMojo {

  @Parameter(defaultValue = "${settings}", readonly = true)
  var settings: Settings = _

  @Parameter(defaultValue = "${project}", readonly = true)
  private var project: MavenProject = _

  override def execute() {
    if (project.getPackaging != "war") {
      getLog.info("Hibernate war patching supports war project.Skip jar/pom projects.")
      return
    }
    var findBeangleHibernate = false
    var findHibernate = false
    val artifactIter = project.getArtifacts.iterator()
    while (artifactIter.hasNext()) {
      val artifact = artifactIter.next
      if (artifact.getArtifactId.startsWith("beangle-data-hibernate")) findBeangleHibernate = true
      if (artifact.getArtifactId == "hibernate-core") findHibernate = true
    }
    if (!(findHibernate && findBeangleHibernate)) return
    try {
      val iter = Patches.files.iterator
      while (iter.hasNext) {
        val patch = iter.next
        val target = project.getBuild.getOutputDirectory + patch
        val is = getClass.getResourceAsStream("/patches" + patch + "file")
        getLog.info("Patching " + patch)
        new File(target.substring(0, target.lastIndexOf("/"))).mkdirs()
        new File(target).createNewFile()
        IOs.copy(is, new FileOutputStream(target))
      }
    } catch {
      case e: Exception => e.printStackTrace()
    } finally {
    }
  }
}
